File: Client-Java/src/main/java/pt/ulisboa/tecnico/tuplespaces/client/grcp/ClientService.java
package pt.ulisboa.tecnico.tuplespaces.client.grcp;

import com.google.protobuf.ProtocolStringList;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.StatusRuntimeException;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesOuterClass.PutRequest;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesOuterClass.ReadRequest;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesOuterClass.ReadResponse;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesOuterClass.TakeRequest;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesOuterClass.TakeResponse;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesOuterClass.getTupleSpacesStateRequest;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesOuterClass.getTupleSpacesStateResponse;
import pt.ulisboa.tecnico.tuplespaces.client.ClientMain;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesGrpc;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesGrpc.TupleSpacesBlockingStub;

import io.grpc.Metadata;
import io.grpc.stub.MetadataUtils;

public class ClientService {
  private ManagedChannel channel;
  private final TupleSpacesBlockingStub stub;
  private final int client_id;

  public ClientService(String host_port, int client_id) {
    // Channel is the abstraction to connect to a service endpoint.
		// Let us use plaintext communication because we do not have certificates.
    this.channel = ManagedChannelBuilder.forTarget(host_port).usePlaintext().build();
    ClientMain.debug(ClientService.class.getSimpleName(), "Client " + client_id + " connected to " + host_port);

		// It is up to the client to determine whether to block the call.
		// Here we create a blocking stub.
    this.stub = TupleSpacesGrpc.newBlockingStub(channel);
    ClientMain.debug(ClientService.class.getSimpleName(), "Client " + client_id + " created a blocking stub");

    this.client_id = client_id;
  }

  private TupleSpacesBlockingStub addMetadataToStub(String[] split) {
    Metadata metadata = new Metadata();
    metadata.put(Metadata.Key.of("delay1", Metadata.ASCII_STRING_MARSHALLER), String.valueOf(split[2]));
    metadata.put(Metadata.Key.of("delay2", Metadata.ASCII_STRING_MARSHALLER), String.valueOf(split[3]));
    metadata.put(Metadata.Key.of("delay3", Metadata.ASCII_STRING_MARSHALLER), String.valueOf(split[4]));
    return MetadataUtils.attachHeaders(stub, metadata);
  }

  // Adds tuple t to the tuple space
  public void put(String[] split) {
    try {
      String tuple = split[1];
      TupleSpacesBlockingStub _stub = this.stub;
      if (split.length == 5) {
        _stub = addMetadataToStub(split);
      }
      ClientMain.debug(ClientService.class.getSimpleName(), "Client " + client_id + " putting tuple " + tuple);
      PutRequest request = PutRequest.newBuilder().setNewTuple(tuple).build();
      _stub.put(request);
      ClientMain.debug(ClientService.class.getSimpleName(), "Client " + client_id + " put tuple " + tuple);
      System.out.println("OK");

    } catch (StatusRuntimeException e) {
      System.out.println("Erro: " + e.getMessage()); // FIXME
      System.out.println("Server is down. Please try again later.");
    }
  }

  // Accepts a tuple description and returns one tuple that matches the description, if it exists.
  // This operation blocks the client until a tuple that satisfies the description exists. The tuple is not removed from the tuple space.
  public String read(String[] split) {
    try {
      String pattern = split[1];
      ClientMain.debug(ClientService.class.getSimpleName(), "Client " + client_id + " reading tuple " + pattern);
      ReadRequest request = ReadRequest.newBuilder().setSearchPattern(pattern).build();
      ReadResponse response = this.stub.read(request);
      ClientMain.debug(ClientService.class.getSimpleName(), "Client " + client_id + " read tuple with pattern " + pattern + " and got " + response.getResult());

      System.out.println("OK");
      return response.getResult();

    } catch (StatusRuntimeException e) {
      System.out.println("Server is down. Please try again later.");
      System.out.println();
      return null;
    }
  }

  // Accepts a tuple description and returns one tuple that matches the description.
  // This operation blocks the client until a tuple that satisfies the description exists. The tuple is removed from the tuple space.
  public String take(String[] split) {
    try {
      String pattern = split[1];
      ClientMain.debug(ClientService.class.getSimpleName(), "Client " + client_id + " taking tuple " + pattern);
      TakeRequest request = TakeRequest.newBuilder().setSearchPattern(pattern).build();
      TakeResponse response = this.stub.take(request);
      ClientMain.debug(ClientService.class.getSimpleName(), "Client " + client_id + " take tuple with pattern " + pattern + " and got " + response.getResult());

      System.out.println("OK");
      return response.getResult();

    } catch (StatusRuntimeException e) {
      System.out.println("Server is down. Please try again later.");
      System.out.println();
      return null;
    }
  }

  // Does not take arguments and returns a list of all tuples on each server.
  public void getTupleSpacesState() {
    try {
      ClientMain.debug(ClientService.class.getSimpleName(), "Client " + client_id + " requested the full tuple space state");
      getTupleSpacesStateRequest request = getTupleSpacesStateRequest.newBuilder().build();
      getTupleSpacesStateResponse response = this.stub.getTupleSpacesState(request);
      ProtocolStringList TupleList = response.getTupleList();

      System.out.println("OK");
      System.out.println(TupleList);
      ClientMain.debug(ClientService.class.getSimpleName(), "Client " + client_id + " received the full tuple space state");

    } catch (StatusRuntimeException e ) {
      System.out.println("Server is down. Please try again later.");
    }
  }

  // A Channel should be shutdown before stopping the process.
  public void shutdown() {
    this.channel.shutdownNow();
  }
}

File: Client-Python/client_service.py
import sys
sys.path.insert(1, '../Contract/target/generated-sources/protobuf/python')

import grpc
from TupleSpaces_pb2 import PutRequest, ReadRequest, ReadResponse, TakeRequest, TakeResponse, getTupleSpacesStateRequest
from TupleSpaces_pb2_grpc import TupleSpacesStub

from debug import Debugger

class ClientService:
    def __init__(self, host_port: str, client_id: int):
        # Create a channel to connect to the server (plaintext communication)
        self.channel = grpc.insecure_channel(host_port)

        # Create a blocking stub for synchronous calls
        self.stub = TupleSpacesStub(self.channel)
        Debugger.debug(f"Client {client_id} connected to the server at {host_port}")

    # Adds tuple t to the tuple space
    def put(self, tuple_str: str):
        Debugger.debug(f"Client {tuple_str} added tuple {tuple_str} to the tuple space")
        request = PutRequest(newTuple=tuple_str)
        self.stub.put(request)
        Debugger.debug(f"Client {tuple_str} added tuple {tuple_str} to the tuple space")

    # Reads a tuple without removing it from the tuple space
    def read(self, pattern: str) -> str:
        Debugger.debug(f"Client {pattern} read tuple {pattern} from the tuple space")
        request = ReadRequest(searchPattern=pattern)
        response: ReadResponse = self.stub.read(request)
        Debugger.debug(f"Client {pattern} read tuple {pattern} from the tuple space")
        return response.result

    # Takes a tuple, removing it from the tuple space
    def take(self, pattern: str) -> str:
        Debugger.debug(f"Client {pattern} took tuple {pattern} from the tuple space")
        request = TakeRequest(searchPattern=pattern)
        Debugger.debug(f"Client {pattern} took tuple {pattern} from the tuple space")
        response: TakeResponse = self.stub.take(request)
        Debugger.debug(f"Client {pattern} took tuple {pattern} from the tuple space")
        return response.result

    # Gets the full tuple space state
    def get_tuple_spaces_state(self):
        Debugger.debug("Client requested the full tuple space state")
        request = getTupleSpacesStateRequest()
        response = self.stub.getTupleSpacesState(request)
        tuples_list = list(response.tuple)

        print("[", end="")
        for i in range(len(tuples_list)):
            tuples_list[i] = tuples_list[i][1:-1]
            print(f"<{tuples_list[i]}>", end="")
            if i != len(tuples_list) - 1:
                print(", ", end="")
        print("]")
        Debugger.debug("Client received the full tuple space state")

    # Shuts down the channel
    def shutdown(self):
        self.channel.close()